akka-http-core/src/main/scala/akka/http/engine/client/HostConnectionPoolConductor.scala

```diff
@@ -0,0 +1,221 @@
+    import FlexiMerge._
+
+    def createMergeLogic(s: FanInShape2[RequestContext, SimpleSlotEvent, SwitchCommand]): MergeLogic[SwitchCommand] =
+      new MergeLogic[SwitchCommand] {
+        val slotStates = Array.fill[SlotState](slotCount)(Unconnected)
+        def initialState = nextState(0)
+        override def initialCompletionHandling = eagerClose
+
+        def nextState(currentSlot: Int): State[_] = {
+          val read: ReadCondition[_] = currentSlot match {
+            case -1 ⇒ Read(s.in1) // if we have no slot available we are not reading from upstream (only SlotEvents)
+            case _  ⇒ ReadAny(s) // otherwise we read SlotEvents *as well as* from upstream
+          }
+          State(read) { (ctx, inlet, element) ⇒
+            element match {
+              case rc: RequestContext ⇒
+                ctx.emit(SwitchCommand(rc, currentSlot))
+                slotStates(currentSlot) = slotStateAfterDispatch(slotStates(currentSlot), rc.request.method)
+              case SlotEvent.RequestCompleted(slotIx) ⇒
+                slotStates(slotIx) = slotStateAfterRequestCompleted(slotStates(slotIx))
+              case SlotEvent.Disconnected(slotIx, failed) ⇒
+                slotStates(slotIx) = slotStateAfterDisconnect(slotStates(slotIx), failed)
+            }
+            nextState(bestSlot())
+          }
+        }
+
+        def slotStateAfterDispatch(slotState: SlotState, method: HttpMethod): SlotState =
+          slotState match {
+            case Unconnected | Idle ⇒ if (method.isIdempotent) Loaded(1) else Busy(1)
+            case Loaded(n)          ⇒ if (method.isIdempotent) Loaded(n + 1) else Busy(n + 1)
+            case Busy(_)            ⇒ throw new IllegalStateException("Request scheduled onto busy connection?")
+          }
+
+        def slotStateAfterRequestCompleted(slotState: SlotState): SlotState =
+          slotState match {
+            case Loaded(1) ⇒ Idle
+            case Loaded(n) ⇒ Loaded(n - 1)
+            case Busy(1)   ⇒ Idle
+            case Busy(n)   ⇒ Busy(n - 1)
+            case _         ⇒ throw new IllegalStateException(s"RequestCompleted on $slotState connection?")
+          }
+
+        def slotStateAfterDisconnect(slotState: SlotState, failed: Int): SlotState =
+          slotState match {
+            case Idle if failed == 0      ⇒ Unconnected
+            case Loaded(n) if n > failed  ⇒ Loaded(n - failed)
+            case Loaded(n) if n == failed ⇒ Unconnected
+            case Busy(n) if n > failed    ⇒ Busy(n - failed)
+            case Busy(n) if n == failed   ⇒ Unconnected
+            case _                        ⇒ throw new IllegalStateException(s"Disconnect(_, $failed) on $slotState connection?")
+          }
+
+        /**
+         * Implements the following Connection Slot selection strategy
+         *  - Select the first idle connection in the pool, if there is one.
+         *  - If none is idle select the first unconnected connection, if there is one.
```

Comment:
Indeed. While it doesn't really matter I'll fix this (easier to follow during debugging).